dCommon/AMFDeserialize.cpp

```diff
@@ -0,0 +1,151 @@
+                double value;
+                inStream->Read(value);
+                doubleValue->SetDoubleValue(value);
+                values->InsertValue(key, doubleValue);
+                break;
+            }
+
+            case AMFValueType::AMFString: {
+                AMFStringValue* stringValue = new AMFStringValue();
+                std::string value;
+                int8_t length;
+                inStream->Read(length);
+                if (length % 2 == 1)  {
+                    length = length >> 1;
+                    for (uint32_t i = 0; i < length; i++) {
+                        int8_t character;
+                        inStream->Read(character);
+                        value.push_back(character);
+                    }
+                    // Empty strings are never sent by reference
+                    if (!value.empty()) accessedElements.push_back(value);
+                } else {
+                    // Length is a reference!
+                    length = length / 2;
+                    value = accessedElements[length];
+                }
+                stringValue->SetStringValue(value);
+                values->InsertValue(key, stringValue);
+                break;
+            }
+
+            case AMFValueType::AMFArray: {                
+                int8_t sizeOfSubArray;
+                int8_t index = 0;
+                inStream->Read(sizeOfSubArray);
+
+                if (sizeOfSubArray % 2 == 1) sizeOfSubArray = sizeOfSubArray >> 1;
+
+                AMFArrayValue* result = new AMFArrayValue();
+                if (sizeOfSubArray >= 1) {
+                    // Need to read this byte!
+                    inStream->Read(index);
+                    for (uint32_t i = 0; i < sizeOfSubArray; i++) {
+                        result->PushBackValue(Read(inStream, true, key));
+                    }
+                } else {
+                    result = Read(inStream, false, key);
+                }
+                values->InsertValue(key, result);
+                break;
+            }
+
+            // Don't think I need these for now.  Will log if I need them.
+            case AMFValueType::AMFInteger:
+            case AMFValueType::AMFXMLDoc:
+            case AMFValueType::AMFDate:
+            case AMFValueType::AMFObject:
+            case AMFValueType::AMFXML:
+            case AMFValueType::AMFByteArray:
+            case AMFValueType::AMFVectorInt:
+            case AMFValueType::AMFVectorUInt:
+            case AMFValueType::AMFVectorDouble:
+            case AMFValueType::AMFVectorObject:
+            case AMFValueType::AMFDictionary:
+                throw static_cast<AMFValueType>(valueType);
+                break;
+        }
+        // Read size of next string
+        inStream->Read(sizeOfString);
+
+        if (sizeOfString % 2 == 0) isReference = true;
+        else isReference = false;
+
+        sizeOfString = sizeOfString >> 1;
```

Comment:
Would it make sense to change the loop to `while (true) { ... break; ... }` to only have this logic for reading and decoding the size once at the start of the loop, instead of it being duplicated before and at the end of the loop here?

Would it maybe also make sense to rename the variable `sizeOfString` to `sizeOfKey` / `sizeOfKeyString` to avoid confusion with a string value?